SafeExamBrowser.SystemComponents/VirtualMachineDetector.cs

```diff
@@ -37,27 +39,136 @@ public VirtualMachineDetector(ILogger logger, ISystemInfo systemInfo)
+			isVirtualMachine |= model.Contains("Q35 +");  // qemu
+
+			return isVirtualMachine;
+		}
+
+		private bool IsVirtualRegistry()
+		{
+			bool isVirtualMachine = false;
+
+			// check historic hardware profiles
+			RegistryKey hardwareConfig = Microsoft.Win32.Registry.LocalMachine.OpenSubKey("SYSTEM\\HardwareConfig");
+
+			if (hardwareConfig != null)
+			{
+				foreach (string configId in hardwareConfig.GetSubKeyNames())
+				{
+					RegistryKey configKey = Microsoft.Win32.Registry.LocalMachine.OpenSubKey($"SYSTEM\\HardwareConfig\\{configId}");
+
+					if (configKey == null)
+					{
+						continue;
+					}
+
+					// reconstruct the systemInfo.biosInfo string
+					string biosInfo = (string) configKey.GetValue("BIOSVendor") + " " + (string) configKey.GetValue("BIOSVersion");
+					string manufacturer = (string) configKey.GetValue("SystemManufacturer");
+					string model = (string) configKey.GetValue("SystemProductName");
+
+					isVirtualMachine |= IsVirtualSystemInfo(biosInfo, manufacturer, model);
+
+					// hardware information of profile throughout installation etc. 
+					RegistryKey computerIds = Microsoft.Win32.Registry.LocalMachine.OpenSubKey($"SYSTEM\\HardwareConfig\\{configId}\\ComputerIds");
+
+					if (computerIds == null)
+					{
+						continue;
+					}
+
+					foreach (string computerId in computerIds.GetSubKeyNames())
+					{
+						// e.g. manufacturer&version&sku&...
+						string computer = (string)computerIds.GetValue(computerId);
+						isVirtualMachine |= IsVirtualSystemInfo(computer, computer, computer);
+					}
+				}
+			}
+
+			// check Windows timeline caches for current hardware config
+			RegistryKey deviceCache = Microsoft.Win32.Registry.CurrentUser.OpenSubKey($"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\TaskFlow\\DeviceCache");
+			
+			if (deviceCache != null)
+			{
+				foreach (string cacheId in deviceCache.GetSubKeyNames())
+				{
+					RegistryKey cacheKey = Microsoft.Win32.Registry.CurrentUser.OpenSubKey($"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\TaskFlow\\DeviceCache\\{cacheId}");
+
+					if (cacheKey == null)
+					{
+						continue;
+					}
+
+					string currHostname = System.Environment.GetEnvironmentVariable("COMPUTERNAME").ToLower();
+					string cacheHostname = ((string) cacheKey.GetValue("DeviceName")).ToLower();
+
+					// windows timeline syncs with other hosts that a user has logged into, hence avoid false positives
+					if (cacheHostname == currHostname)
+					{
+						string biosInfo = "";
+						string manufacturer = (string) cacheKey.GetValue("DeviceMake");
+						string model = (string) cacheKey.GetValue("DeviceModel");
+
+						isVirtualMachine |= IsVirtualSystemInfo(biosInfo, manufacturer, model);
+					}
+				}
+			}
+
+			return isVirtualMachine;
+		}
+
+		private bool IsVirtualWmi()
+		{
+			bool isVirtualMachine = false;
+
+			ManagementObjectSearcher searcherCpu = new ManagementObjectSearcher("SELECT * FROM Win32_Processor");
```

Comment:
This WMI code should be placed into a try-catch and then moved to `ISystemInfo`, where there already are other related WMI queries.